SDAAP-101: Fix issue with find pattern matching replace

rbi-aap opened the pull request:
- Resolved problem where find pattern incorrectly matched the replace
- Improved regex handling to ensure correct replacements
- Added test cases to verify expected behavior

rbi-aap committed:
https://sofab.atlassian.net/browse/SDAAP-101,

rbi-aap committed:
SDAAP-101: Fix issue with find pattern matching replace
- Resolved problem where find pattern incorrectly matched the replace
- Improved regex handling to ensure correct replacements
- Added test cases to verify expected behavior

rbi-aap commented:
Hi Andrew and Team, please review, thanks

MarkLark86 commented:
@tomaskikutis Could you please review this PR, and see if it is also needed in v2.8, 2.9 & develop branches please.
Cheers

tomaskikutis commented:
@rbi-aap if the browser was freezing it was probably due to infinite loop. Why was it causing an infinite loop?

Is your code only handling the infinite loop issue? If not, what other issues is it handling?

I see after your changes, 2 different functions exist - one to replace at index and another to replace all. Could you stick to a single implementation in a similar fashion how it was previously? e.g. when multiple replacements were needed, the function doing a single replacement would be called multiple times.

rbi-aap committed:
update the test description detail

rbi-aap committed:
Revert "update the test description detail"

This reverts commit 2b777164306430c3696e84a35f9cdc5d9264d918.

rbi-aap commented:
@tomaskikutis, 
Thanks for the review. I refactored the replacement logic to improve stability, prevent unintended behavior, and ensure correct text updates. Unit tests were added to validate edge cases. Appreciate your time.

tomaskikutis reviewed (changes requested):
Could you answer to each of my 3 questions separatelly?

rbi-aap commented:
@tomaskikutis ,Thanks for the review. The infinite loop was caused by unintended offset shifts leading to continuous execution and in turn, possible issues like callback hell, which have been fixed for stability. The update also prevents duplicate replacements and ensures correct text updates. The functions were split to avoid instability, as a single implementation didn’t handle multiple replacements reliably. thanks

rbi-aap committed:
add more test case

rbi-aap committed:
Fix merge conflicts in reducers.spec.tsx and add more reducer test cases

- Resolved merge conflicts between HEAD and commit 59502011c909e4a2a8e3b435fbf9877c2657851b
- Merged duplicate test cases while keeping the most relevant details
- Added additional test cases for:
  - HIGHLIGHTS_REPLACE_ALL edge cases
  - Handling special characters in search/replace
- Improved assertions to ensure coverage for different scenarios

rbi-aap commented:
Hi Team, can you review more test case for QA with its function, thanks

rbi-aap committed:
update for combine single and all replacement from pattern

tomaskikutis commented:
Can you provide a sample input text, pattern and replacement text which was causing the infinite loop?

> The update also prevents duplicate replacements and ensures correct text updates.

Can you specify samples where it was causing duplicate replacements or incorrect updates?

<hr>

Please rework the tests so they are simpler
- limit sample text to a single line.
- Clean up test names so it is clear which use case is being tested. For example - this it not clear at all - "HIGHLIGHTS_FIND_REPLACE_SINGLE_AND_REPLACE_ALL"

<hr>

`Replace all $ with $AUD ,but no change with $AUD` - this one is asserting incorrect behavior. It should replace `$AUD` too with the pattern that you specified.

```javascript
'I have $100 in my wallet. The total cost is $50, but sometimes it is $AUD60'.replace(/\$/g, '$AUD')
// > 'I have $AUD100 in my wallet. The total cost is $AUD50, but sometimes it is $AUDAUD60'
```

rbi-aap committed:
Add organised test suite for highlight and replace functionality
- Structured tests into Single, All, and Combined sections
- Improved readability with Arrange-Act-Asse pattern and descriptive names

rbi-aap commented:
Hi @tomaskikutis, 

Thanks for your review!

- Infinite loop sample: "$100" with pattern "$" and replacement "$AUD" looped endlessly.
- Update resolves duplicates and ensures correct replacements.
- Fixed duplicate sample: "$AUD60" no longer becomes "$AUDAUD60" with "$" → "$AUD".
- Tests are simpler, using AAA pattern—see new commit.
- Sample text now single-line, e.g., "Total: $100, cost: $50".
- Test names improved, e.g., "CombinedReplacementScenarios".
- Added QA test for edge cases.

Really appreciate your support!

tomaskikutis commented:
I've looked into the issue and indeed there is a bug with an infinite loop. The cause was that the function was performing replacement on text that was already replaced.

For example if text was `air` and we wanted to replace "air" with "airplane", it would do it correctly the first iteration, but then would search for "air" in "airplane" and would replace to "airplaneplane", then to "airplaneplaneplane" - and would never stop.

I've fixed the infinite loop in https://github.com/superdesk/superdesk-client-core/pull/4758 by adjusting code so it does not attempt to replace what was already replaced once.

Besides that - I don't see any other issues.

Your comment about duplicates is describing correct behavior. I've added a test that given text "dev dev dev" and trying to replace "dev" with "developers" it should output "developers developers developers". It's not a bug - that's how "replace all" is supposed to work.

Regarding currency conversion where you want to change "$" to "$AUD" - but keep existing mentions of "$AUD" intact - that can't be done with "find and replace" feature without regex support. I don't suggest to expose regex support directly to end-users either - that'd be an overkill for superdesk users.

The best solution I think would be to have pre-made custom replacement strategies. It could either be integrated into "find and replace" widget or you could use macros. There is one already for [USD to CAD](https://github.com/user-attachments/assets/c4f879e2-2da2-4a3a-9b18-47147bd31476) conversion.

rbi-aap committed:
fix infinite loop, shrinking bug & improve $ ↔ $AUD replacements like pattern

- Fixed shrinking bug ($AUD → $) to prevent skipped replacements
- Improved regex handling to prevent infinite loops & incorrect lastIndex resets, like $->$AUD pattern
- Enhanced $ → $AUD replacement to avoid duplicate matches
- Ensured correct index updates to maintain text integrity
- Improved table entity detection & handling for replacements in tables
- others

rbi-aap commented:
Hi Team, 
Please review the updated PR
### Fix infinite loop, shrinking bug & improve find-replace logic
### Summary
This PR addresses key issues in find-replace functionality:
- Fixed shrinking replacement issue ($AUD → $) to ensure correct replacements.
- Prevented infinite loop by correctly updating `lastIndex` in regex.
- Improved&FIX  handling of $ → $AUD likewise pattern
- Enhanced table entity detection for find-replace inside tables.
### 
QA & Testing
- Unit tests cover shrinking & expanding replacements.
- Tested `$ → $AUD` and `$AUD → $` transformations for consistency.
- able replacement logic verified.

#Review Notes
- [ ]  Please check if all edge cases are correctly handled.
- [ ] Let me know if you find any further improvements.
Thanks for reviewing.

tomaskikutis commented:
I've merged the infinite loop fix (https://github.com/superdesk/superdesk-client-core/pull/4758) and cherry-picked to [hotfix/2.5.3](https://github.com/superdesk/superdesk-client-core/tree/hotfix/2.5.3).

Besides that, I'm not aware of any other issues. As I've mentioned, your comments about "duplicates" are describing correct behavior.

Let me know if you see any legitimate bugs remaining.
